lib/SPIRV/libSPIRV/SPIRVBasicBlock.cpp

```diff
@@ -58,15 +58,37 @@ SPIRVDecoder SPIRVBasicBlock::getDecoder(std::istream &IS) {
   return SPIRVDecoder(IS, *this);
 }
 
+/// If insertion of a new instruction before the one passed to the function
+/// is illegal, insertion before the returned instruction is guaranteed
+/// to retain correct instruction order in a block
+SPIRVInstruction *SPIRVBasicBlock::getVerifiedInsertBefore(
+    const SPIRVInstruction *RequestedInsertBefore) {
+  if (RequestedInsertBefore) {
+    switch (RequestedInsertBefore->getOpCode()) {
+    case OpBranch:
+    case OpBranchConditional: {
+      auto Prev = RequestedInsertBefore->getPrevious();
+      if (Prev && Prev->getOpCode() == OpLoopMerge)
+        return Prev;
+      break;
+    }
+    default:
+      break;
+    }
+  }
+  return const_cast<SPIRVInstruction *>(RequestedInsertBefore);
+}
+
 /// Assume I contains valid Id.
 SPIRVInstruction *
 SPIRVBasicBlock::addInstruction(SPIRVInstruction *I,
                                 const SPIRVInstruction *InsertBefore) {
   assert(I && "Invalid instruction");
   Module->add(I);
   I->setParent(this);
-  if (InsertBefore) {
-    auto Pos = std::find(InstVec.begin(), InstVec.end(), InsertBefore);
```

Comment:
I think it was much easier if the following piece of code was added here.
```
    if (isa<OpLoopMerge>((*Pos)->getPrevious()))
      --Pos;
```
Also meaningfull comment would be helpful.

Reply:
It would be much easier indeed :) Done likewise